### PR TITLE
tests/raftstore: fix CI

### PR DESCRIPTION
**Pull request by tiancaiamao**
drop packet test case should be fixed, almostly, if not absolutely
some other problem in CI 'may' also be fixed

### CODE REVIEW

File: tests/raftstore/cluster.rs [hunk 1]
```diff
@@ -313,23 +315,34 @@ impl<T: Simulator> Cluster<T> {
...
+                        } else if try_cnt == 1 && resp.get_header().get_error().has_stale_epoch() {
+                            warn!("seems split, let's retry");
+                            try_cnt += 1;
+                            continue;
+                        }
+                    }
+                    return resp;
+                }
+                Err(e) => {
+                    if let Error::Timeout(_) = e {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: node and server `call_command` doesn't return this timeout error. how do you check?

File: tests/raftstore/cluster.rs [hunk 2]
```diff
@@ -313,23 +315,34 @@ impl<T: Simulator> Cluster<T> {
...
+                        warn!("call command timeout, let's retry");
+                        continue;
+                    }
+                    assert!(false);
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: panic with detailed resp information.

File: tests/raftstore/cluster.rs [hunk 3]
```diff
@@ -297,7 +298,8 @@ impl<T: Simulator> Cluster<T> {
 
         let err = resp.get_header().get_error();
         if !err.has_not_leader() {
-            return false;
+            self.reset_leader_of_region(region_id);
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: the error is not "NotLeader", maybe other like stale epoch, so why reset leader?

File: tests/raftstore/cluster.rs [hunk 4]
```diff
@@ -142,21 +142,22 @@ impl<T: Simulator> Cluster<T> {
...
     pub fn call_command_on_leader(&mut self,
                                   region_id: u64,
                                   mut request: RaftCmdRequest,
                                   timeout: Duration)
                                   -> Result<RaftCmdResponse> {
-        let leader = self.leader_of_region(region_id).unwrap();
-        request.mut_header().set_peer(leader);
-        self.call_command(request, timeout)
+        loop {
+            if let Some(leader) = self.leader_of_region(region_id) {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: don't loop forever if we can't get the leader at all.

File: tests/raftstore/cluster.rs [hunk 5]
```diff
@@ -313,23 +315,34 @@ impl<T: Simulator> Cluster<T> {
...
-                    try_cnt += 1;
-                    continue;
+            let result = self.call_command_on_leader(region_id, req, timeout);
+            match result {
+                Ok(resp) => {
+                    if resp.get_header().has_error() {
+                        if self.refresh_leader_if_needed(&resp, region_id) {
+                            warn!("seems leader changed, let's retry");
+                            continue;
+                        } else if try_cnt == 1 && resp.get_header().get_error().has_stale_epoch() {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: why` try_cnt == 1`?

File: tests/raftstore/cluster.rs (hunk 1 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: that's because we map it to Other(String) error previously...I've remove that line

        .map_err(|e| box_err!("call command failed {:?}, cost {:?}", e, t.elapsed()))

Comment (siddontang) [inline tests/raftstore/cluster.rs]: no, node channel call_command and server tcp read both return other timeout error.

Comment (siddontang) [inline tests/raftstore/cluster.rs]: you can't only check with your own strategy.

Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: Well, if we met a Timeout kind of error, whatever it is, retry... 
if it's not a Timeout, we can't cope with it yet

File: tests/raftstore/cluster.rs (hunk 3 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: reset the cached leader in Cluster's, so next time query goes to Pd, chance that we'll get the right leader.

Comment (siddontang) [inline tests/raftstore/cluster.rs]: why not leader doesn't reset but others do? 
If we get stale epoch error, the peer is still leader, why reset?

**Comment by tiancaiamao** [comment]
PTAL @BusyJay @siddontang @ngaut

**Comment by ngaut** [comment]
LGTM, Good job.

File: tests/raftstore/cluster.rs [hunk 6]
```diff
@@ -142,21 +142,33 @@ impl<T: Simulator> Cluster<T> {
...
     pub fn call_command_on_leader(&mut self,
                                   region_id: u64,
                                   mut request: RaftCmdRequest,
                                   timeout: Duration)
                                   -> Result<RaftCmdResponse> {
-        let leader = self.leader_of_region(region_id).unwrap();
-        request.mut_header().set_peer(leader);
-        self.call_command(request, timeout)
+        let mut retry = 0;
+        loop {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: `for _ in 0..200` instead loop

File: tests/raftstore/cluster.rs [hunk 7]
```diff
@@ -142,21 +142,33 @@ impl<T: Simulator> Cluster<T> {
...
+            match self.leader_of_region(region_id) {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: use `if let Some()`

File: tests/raftstore/cluster.rs [hunk 8]
```diff
@@ -142,21 +142,33 @@ impl<T: Simulator> Cluster<T> {
...
+                Some(leader) => {
+                    request.mut_header().set_peer(leader);
+                    return self.call_command(request, timeout);
+                }
+                None => {
+                    retry += 1;
+                    if retry > 200 {
+                        return Err(Error::Timeout("can't get leader of region after retry 200 \
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: return this error after the for loop

File: tests/raftstore/cluster.rs (hunk 7 above)
Comment (siddontang) [inline tests/raftstore/cluster.rs]: do some sleep if leader not found .

File: tests/raftstore/cluster.rs [hunk 9]
```diff
@@ -318,18 +331,29 @@ impl<T: Simulator> Cluster<T> {
...
+                            continue;
+                        } else if try_cnt <= 5 && resp.get_header().get_error().has_stale_epoch() {
+                            warn!("seems split, let's retry");
+                            try_cnt += 1;
+                            continue;
+                        }
+                    }
+                    return resp;
+                }
+                Err(e) => {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]:
```
Err(Error::Timeout(_)) => {}
Err(e) => panic
```

**Comment by tiancaiamao** [comment]
PTAL @siddontang

File: tests/raftstore/cluster.rs [hunk 10]
```diff
@@ -318,18 +322,27 @@ impl<T: Simulator> Cluster<T> {
...
+                            continue;
+                        } else if try_cnt <= 5 && resp.get_header().get_error().has_stale_epoch() {
+                            warn!("seems split, let's retry");
+                            try_cnt += 1;
+                            continue;
+                        }
+                    }
+                    return resp;
+                }
+                Err(Error::Timeout(_)) => {
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: can't loop forever

Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: it won't loop forever.
if resp is ok (but resp.get_header() may has error), or we met a error other than Timeout, loop will break

Comment (siddontang) [inline tests/raftstore/cluster.rs]: if timeout every time?

Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: it timeout always happen, the test cases won't pass... **STILL** we'll back again and fix CI
so...seems that it don't matter loop here

Comment (BusyJay) [inline tests/raftstore/cluster.rs]: How do we know that it's safe to just retry when timeout?

Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: @BusyJay 
Some operations are not idempotent and we can't simply retry when timeout.
we still have the issue here https://github.com/pingcap/tikv/issues/427 to be solved.

The purpose of this PR is try to fix CI of drop packet case,  based on the observation:
call command on a stale leader, or leader not elected out, request will timeout.
then our test code will unwrap on None or Err and panic. 
on both situation retry is safe

it don't solve the issue427, it just works and benefit some other tests in CI.

File: tests/raftstore/cluster.rs [hunk 11]
```diff
@@ -318,18 +322,27 @@ impl<T: Simulator> Cluster<T> {
...
+                    warn!("call command timeout, let's retry");
```
Comment (ngaut) [inline tests/raftstore/cluster.rs]: We may just retry read-only command

File: tests/raftstore/cluster.rs (hunk 10 above)
Comment (BusyJay) [inline tests/raftstore/cluster.rs]: Sounds like a better solution is to make stale leader return not leader error rather than just retry.

File: tests/raftstore/cluster.rs (hunk 11 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: It's two kind of thing:
one is we know exactly that timeout reason caused by "current cluster has no leader" or "we send request to stale leader, which don't reply because it's not leader any more "
the other is we don't know what cause timeout but we have a idempotent operation, it's okay to retry.

retry the former case fix CI but not always right (if timeout by other reason)
retry the later case always do things right but never fix anything

Comment (siddontang) [inline tests/raftstore/cluster.rs]: Now our tests use put, I think it is safe to retry.

File: tests/raftstore/cluster.rs (hunk 10 above)
Comment (siddontang) [inline tests/raftstore/cluster.rs]: we may still handle timeout here, especially for later partition strategy.
So we should both doing 1) calling callback when lost leader, 2) retry if possible for timeout.

Comment (siddontang) [inline tests/raftstore/cluster.rs]: > it timeout always happen, the test cases won't pass... STILL we'll back again and fix CI
so...seems that it don't matter loop here

Please consider all corner cases, if this happen, CI will wait for a long time and then terminate.

**Comment by siddontang** [comment]
LGTM 

PTAL @ngaut @BusyJay

File: tests/raftstore/cluster.rs (hunk 3 above)
Comment (siddontang) [inline tests/raftstore/cluster.rs]: why reset here?

File: tests/raftstore/cluster.rs [hunk 12]
```diff
@@ -142,21 +142,24 @@ impl<T: Simulator> Cluster<T> {
...
 
     pub fn call_command_on_leader(&mut self,
                                   region_id: u64,
                                   mut request: RaftCmdRequest,
                                   timeout: Duration)
                                   -> Result<RaftCmdResponse> {
-        let leader = self.leader_of_region(region_id).unwrap();
-        request.mut_header().set_peer(leader);
-        self.call_command(request, timeout)
+        for _ in 0..200 {
```
Comment (BusyJay) [inline tests/raftstore/cluster.rs]: Why do we need a for loop here?

File: tests/raftstore/cluster.rs [hunk 13]
```diff
@@ -313,24 +317,32 @@ impl<T: Simulator> Cluster<T> {
...
-            let resp = self.call_command_on_leader(region_id, req, timeout).unwrap();
-            if resp.get_header().has_error() {
-                if self.refresh_leader_if_needed(&resp, region_id) {
-                    warn!("seems leader changed, let's retry");
-                    continue;
-                } else if try_cnt == 1 && resp.get_header().get_error().has_stale_epoch() {
-                    warn!("seems split, let's retry");
-                    try_cnt += 1;
+            let result = self.call_command_on_leader(region_id, req, timeout);
+            match result {
```
Comment (BusyJay) [inline tests/raftstore/cluster.rs]: Use `if let` to reduce indent.

**Comment by BusyJay** [comment]
rest LGTM

File: tests/raftstore/cluster.rs (hunk 12 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: leader_of_region may return None, but we want call_command_on_leader
so we retry many times...even we get leader and call command on it, we may still retry many times because the leader we get may be stale

or simply explain: we retry to death!

Comment (BusyJay) [inline tests/raftstore/cluster.rs]: I don't think it's a good idea. If there is a problem, we need to expose it earlier rather than just blindly retry.

In this certain case we should increase the retry count in `leader_of_region` rather than wrap it in another loop.

File: tests/raftstore/cluster.rs (hunk 13 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: it's a bit complex here, I want to handle Ok and some of Err case.

is that looks like:

        if let Ok(xxx) = result {
        }
        if let Err(e) = result {
                if let Error::Timeout(xxx) = e {
                }
        }

will compiler complain result moved?

Comment (BusyJay) [inline tests/raftstore/cluster.rs]:
```
if let Err(Timeout(...)) = res {
   continue
}
let res = res.unwrap();
```

File: tests/raftstore/cluster.rs (hunk 12 above)
Comment (tiancaiamao) [inline tests/raftstore/cluster.rs]: I agree. 
but we have already tons of code call_command_on_leader().unwrap() like patten, make CI vulnerable. we should clean up code in following PRs...
